File: bfalgo/regex/nfa.py
import sys

def regex_to_postfix(re: str) -> str:
    """Convert RegExp to postfix expressions. Eg.
        eg.1: (ab)?ba -> ab.?b.a.
        eg.2: a(bb)+a -> abb.+.a.
        eg.3: a(bb)?c*b|abc -> abb.?.c*.b.ab.c.|
        eg.4: a(b|c)c*a+|abc|a(a+) -> abc|.c*.a+.ab.c.aa+.||

    Construct postfix string according to RegEx's operator precedence:
        [] > () > *+? > concat > ^$ > |

    n_atom counts the atomic operands, such as (ab) in eg.1, (bb) in eg.2,
     c* in eg.3, and all single symbolacters without trailing meta-symbols (
         like b, a in eg.1). Whenever n_atom > 1, we add a dot(.) to dst.

    n_alt counts the number of "|" (alternatives). Because alternative operator
    has the lowest precedence, we make sure characters on the two sides of |
    NOT concatenate by concatenate everything on the left ("clearance").

    Args:
        re (str): [description]

    Returns:
        str: [description]
    """
    dst = ''
    n_alt = 0
    n_atom = 0
    idx_p = 0
    p = []
    for symbol in re:
        if symbol == '(':
            if n_atom > 1:
                n_atom -= 1
                dst = dst + '.'

            # Save the snapshot before entering ()
            p.append({'n_alt': n_alt, 'n_atom': n_atom})
            n_atom = 0
            n_alt = 0
        elif symbol == ')':
            if len(p) == 0:
                return None
            if n_atom == 0:
                return None
            # n_atom should be at least 1
            while n_atom > 1:  
                n_atom -= 1
                dst = dst + '.'
            while n_alt > 0:
                n_alt -= 1
                dst = dst + '|'
            curr_paren = p.pop()

            # Restore the snapshot before entering on exiting.
            n_alt = curr_paren['n_alt']
            n_atom = curr_paren['n_atom']
            n_atom += 1
        elif symbol == '|':
            if n_atom == 0:
                return None
            while n_atom > 1:
                n_atom -= 1
                dst = dst + '.'
            # Force clear n_atom, no concatenation between the two sides.
            n_atom = 0  
            n_alt += 1
        elif symbol in ['*', '+', '?']:
            # special symbols cannot exist without normal chars.
            if n_atom == 0:
                return None
            dst = dst + symbol
        else:
            if n_atom > 1:
                n_atom -= 1
                dst = dst + '.'
            dst = dst + symbol
            n_atom += 1

    if len(p) != 0:
        return None
    while n_atom > 1:
        n_atom -= 1
        dst = dst + '.'
    while n_alt > 0:
        n_alt -= 1
        dst = dst + '|'

    return dst

# for ascii characters
SPLIT = 256  
MATCHED = 257
state_id = 0

class State:
    def __init__(self, value, out, out1):
        """
        State is classified by value,
            value (int): 
                <256 ('consume'), 
                256 ('split'), 
                257 ('stop' or 'matched') 
        and has different out pointers:
            out (State): 
            out1 (State): 
        """
        global state_id
        self.state_id = state_id
        self.value = value
        self.out = out  # a single state or None
        self.out1 = out1  # a single state or None
        state_id += 1

    def collect_state(s, layer, all_states):
        # !!!Danger: mind the loops
        if s.out is not None:
            collect_state(s.out, layer+1)
        if s.out1 is not None:
            collect_state(s.out1, layer+1)
        
    def __repr__(self):
        collect_state(self, 0)

    

class Fragment:
    """An NFA is composed of Fragment's. Each fragment is a black box,
    exposing to outside world with a start state and a list of connected
    pointers, pointing to None or some state. 
    """
    def __init__(self, start: State, out: 'list[State]'):
        self.start = start
        self.out = out

def patch(outlist, state):
    """

    Args:
        outlist ([type]): [description]
        state ([type]): [description]
    """
    outlist.pop()
    outlist.append(state)
    return outlist

def join_outs(outlist1: 'list[State]', outlist2: 'list[State]'):
    """The last element of outlist1 is always None

    Args:
        outlist1 (List[State]): [description]
        outlist2 (List[State]): [description]
    """
    outlist1.pop()
    return outlist1.extend(outlist2)

def list1(state: State):
    return [state] 

def postfix_to_nfa(postfix):
    nfa_nodes = []
    for rc in postfix:
        if rc == '.':
            e2 = nfa_nodes.pop()
            e1 = nfa_nodes.pop()
            patch(e1.out, e2.start)
            nfa_nodes.append(Fragment(e1.start, e2.out))
        elif rc == '|':
            e2 = nfa_nodes.pop()
            e1 = nfa_nodes.pop()
            s = State(SPLIT, e1.start, e2.start)
            nfa_nodes.append(Fragment(s, join_outs(e1.out, e2.out)))
        elif rc == '*':
            e1 = nfa_nodes.pop()
            s = State(SPLIT, e1.start, None)
            patch(e1.out, s)
            nfa_nodes.append(Fragment(s, list1(s.out1)))
        elif rc == '?':
            e1 = nfa_nodes.pop()
            s = State(SPLIT, e1.start, None)
            nfa_nodes.append(Fragment(s, join_outs(e1.out, list1(s.out1))))
        elif rc == '+':
            e1 = nfa_nodes.pop()
            s = State(SPLIT, e1.start, None)
            patch(e1.out, s)
            nfa_nodes.append(Fragment(e1.start, list1(s.out1)))
        else:
            s = State(rc, None, None)
            nfa_nodes.append(Fragment(s, list1(s.out)))
    
    e = nfa_nodes.pop()
    if len(nfa_nodes) != 0:
        return None
    patch(e.out, State(MATCHED, None, None))
    return e

def ismatch(l: 'list[State]') -> bool:
    for s in l:
        if s.value == MATCHED:
            return 1
    return 0

def addstate(state_list: 'list[State]', state: State):
    if state.value == SPLIT:
        addstate(state_list, state.out)
        addstate(state_list, state.out1)
        return    
    else:
        state_list.append(state)

def step(current_states, c):
    next_states = []
    for s in current_states:
        if s.value == c:
            addstate(next_states, s.out) 
    return next_states 

def match(start: State, string: str) -> bool:
    clist = [start]
    nlist = []
    for c in string:
        c = ord(c)
        clist = step(clist, c)
    return ismatch(clist)

def main(reg, string):
    postfix = regex_to_postfix(reg)
    if postfix is None:
        print(f"Bad regexp {reg}\n")

    nfa = postfix_to_nfa(postfix)
    print("ho")

    if match(nfa, string):
        print(f"{string} matched RegEx {reg}")
    return 0

if __name__ == "__main__":
    test_cases = [
        ("(ab)?ba", "ab.?b.a."), 
        ("a(bb)+a", "abb.+.a."),
        ("a(bb)?c*b|abc", "abb.?.c*.b.ab.c.|"), 
        ("a(b|c)c*a+|abc|a(a+)", "abc|.c*.a+.ab.c.aa+.||")
        ]
    for reg, s in test_cases:
        if regex_to_postfix(reg) != s:
            print(f"{reg} --> {regex_to_postfix(reg)} (should be {s})")

    reg, string = sys.argv[1:]
    main(reg, string)

File: bfalgo/regex/nfa.c
/*
 * Regular expression implementation.
 * Supports only ( | ) * + ?.  No escapes.
 * Compiles to NFA and then simulates NFA
 * using Thompson's algorithm.
 *
 * See also http://swtch.com/~rsc/regexp/ and
 * Thompson, Ken.  Regular Expression Search Algorithm,
 * Communications of the ACM 11(6) (June 1968), pp. 419-422.
 * 
 * Copyright (c) 2007 Russ Cox.
 * Can be distributed under the MIT license, see bottom of file.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/*
 * Convert infix regexp re to postfix notation.
 * Insert . as explicit concatenation operator.
 * Cheesy parser, return static buffer.
 */
char* re2post(char *re) { 

	int nalt, natom;
    // Why static variable, when it's called only once? 
    // buf[] will be returned for use in other parts, so must be 
    // defined statically.
	static char buf[8000];
	char *dst;
    // paren for parenthesis
	struct {
		int nalt;
		int natom;
	} paren[100], *p;
	
	p = paren;
	dst = buf;
	nalt = 0;
	natom = 0;
	if(strlen(re) >= sizeof buf/2)
		return NULL;
	for(; *re; re++){
		switch(*re){
		case '(':
			if(natom > 1){
				--natom;
				*dst++ = '.';
			}
			if(p >= paren+100)
				return NULL;
			p->nalt = nalt;
			p->natom = natom;
			p++;
			nalt = 0;
			natom = 0;
			break;
		case '|':
			if(natom == 0)
				return NULL;
			while(--natom > 0)
				*dst++ = '.';
			nalt++;
			break;
		case ')':
			if(p == paren)
				return NULL;
			if(natom == 0)
				return NULL;
			while(--natom > 0)
				*dst++ = '.';
			for(; nalt > 0; nalt--)
				*dst++ = '|';
			--p;
			nalt = p->nalt;
			natom = p->natom;
			natom++;
			break;
		case '*':
		case '+':
		case '?':
			if(natom == 0)
				return NULL;
			*dst++ = *re;
			break;
		default:
			if(natom > 1){
				--natom;
				*dst++ = '.';
			}
			*dst++ = *re;
			natom++;
			break;
		}
	}
	if(p != paren)
		return NULL;
	while(--natom > 0)
		*dst++ = '.';
	for(; nalt > 0; nalt--)
		*dst++ = '|';
	*dst = 0;
	return buf;
}

/*
 * Represents an NFA state plus zero or one or two arrows exiting.
 * if c == Match, no arrows out; matching state.
 * If c == Split, unlabeled arrows to out and out1 (if != NULL).
 * If c < 256, labeled arrow with character c to out.
 */
enum {
	Match = 256,
	Split = 257
};
typedef struct State State;
struct State {
	int c;
	State *out;
	State *out1;
	int lastlist;
};
State matchstate = { Match };	/* matching state */
int nstate; // each state has an id, but it's not used, only as counter

/* Allocate and initialize State */
State* state(int c, State *out, State *out1) {
	State *s;
	
	nstate++;
	s = malloc(sizeof *s);
	s->lastlist = 0;
	s->c = c;
	s->out = out;
	s->out1 = out1;
	return s;
}

/*
 * A partially built NFA without the matching state filled in.
 * Frag.start points at the start state.
 * Frag.out is a list of places that need to be set to the
 * next state for this fragment.
 */
typedef struct Frag Frag;
typedef union Ptrlist Ptrlist;
struct Frag {
	State *start;
	Ptrlist *out;
};

/* Initialize Frag struct. */
Frag frag(State *start, Ptrlist *out) {
	Frag n = { start, out };
	return n;
}

union Ptrlist {
	/*
	A usual node of a linked list, except that the data and the next-pointing
	pointer share the same memory segment. 
	* Since the out pointers in the list are always 
	* uninitialized, we use the pointers themselves
	* as storage for the Ptrlists.
	*/
	Ptrlist *next;
	State *s;
};

Ptrlist* list1(State **outp) {
	/* 
	Create singleton list containing just outp. 

	Args:
		**outp is a pointer to pointer that points to outp, which is a state.

	A ptrlist is an array of states / pointers. 
	A ptrlist pointer is a pointer pointing to this array.
	*/
	Ptrlist *l;
	
	//Make outp a pointer to Ptrlist (which is indeed a pointer to state)
	//(*l).next = Null --> that will erase (*l).s, doesn't that matter? --
	//No. Each time list1 is called, outp is NULL.
	l = (Ptrlist*)outp;
	l->next = NULL;
	return l;
}

void patch(Ptrlist *l, State *s) {
	/* 
	* Patch the list of states at out (of an NFA fragment) to point to start. 
	*/
	Ptrlist *next;
	
	for(; l; l=next){
		/* 
		Starting from head of *l (linked list).
		l->next, l->s are two pointers that can overwrite each other.
		Here we save l->next (can be NULL, or some State)
		*/
		next = l->next;
		l->s = s;
	}
}

Ptrlist* append(Ptrlist *l1, Ptrlist *l2) {
	/* 
	* Join the two lists l1 and l2, returning the combination. 
	*/
	Ptrlist *oldl1;
	
	oldl1 = l1;
	while(l1->next)
		l1 = l1->next;
	l1->next = l2;
	return oldl1;
}

State* post2nfa(char *postfix) {
	/*
	* Convert postfix regular expression to NFA.
	* Return start state.
	*/
	char *p;
	Frag stack[1000], *stackp, e1, e2, e;
	State *s;
	
	printf("postfix: %s\n", postfix);

	if(postfix == NULL)
		return NULL;

	#define push(s) *stackp++ = s
	#define pop() *--stackp

	stackp = stack;
	for(p=postfix; *p; p++){
		switch(*p){
		default:
			s = state(*p, NULL, NULL);
			// precedence [->] > [&], so it's like &(s->out)
			push(frag(s, list1(&s->out)));
			break;
		case '.':	/* catenate */
			e2 = pop();
			e1 = pop();
			patch(e1.out, e2.start);
			push(frag(e1.start, e2.out));
			break;
		case '|':	/* alternate */
			e2 = pop();
			e1 = pop();
			s = state(Split, e1.start, e2.start);
			push(frag(s, append(e1.out, e2.out)));
			break;
		case '?':	/* zero or one */
			e = pop();
			s = state(Split, e.start, NULL);
			push(frag(s, append(e.out, list1(&s->out1))));
			break;
		case '*':	/* zero or more */
			e = pop();
			s = state(Split, e.start, NULL);
			patch(e.out, s);
			push(frag(s, list1(&s->out1)));
			break;
		case '+':	/* one or more */
			e = pop();
			s = state(Split, e.start, NULL);
			patch(e.out, s);
			push(frag(e.start, list1(&s->out1)));
			break;
		}
	}

	e = pop();
	if(stackp != stack)
		return NULL;

	patch(e.out, &matchstate);
	return e.start;
#undef pop
#undef push
}

typedef struct List List;
struct List
{
	State **s;
	int n;
};
List l1, l2;
static int listid;

void addstate(List*, State*);
void step(List*, int, List*);

/* Compute initial state list */
List* startlist(State *start, List *l) {
	l->n = 0;
	listid++;
	addstate(l, start);
	return l;
}

/* Check whether state list contains a match. */
int ismatch(List *l) {
	int i;

	for(i=0; i<l->n; i++)
		if(l->s[i] == &matchstate)
			return 1;
	return 0;
}

/* Add s to l, following unlabeled arrows. */
void addstate(List *l, State *s) {
	if(s == NULL || s->lastlist == listid)
		return;
	s->lastlist = listid;
	if(s->c == Split) {
		/* follow unlabeled arrows */
		addstate(l, s->out);
		addstate(l, s->out1);
		return;
	}
	l->s[l->n++] = s;
}

void step(List *clist, int c, List *nlist) {
	/*
	* Step the NFA from the states in clist
	* past the character c,
	* to create next NFA state set nlist.
	*/
	int i;
	State *s;

	listid++;
	nlist->n = 0;
	for(i=0; i<clist->n; i++){
		s = clist->s[i];
		// why is 's->out1' omitted?
		// because split-state is never added to clist
		// all are single outs
		if(s->c == c)
			addstate(nlist, s->out);
	}
}

/* Run NFA to determine whether it matches s. */
int match(State *start, char *s) {
	int i, c;
	List *clist, *nlist, *t;

	clist = startlist(start, &l1);
	nlist = &l2;
	for(; *s; s++){
		c = *s & 0xFF;
		step(clist, c, nlist);
		t = clist; clist = nlist; nlist = t;	/* swap clist, nlist */
	}
	return ismatch(clist);
}

int main(int argc, char **argv) {
	int i;
	char *post;
	State *start;

	if(argc < 3){
		fprintf(stderr, "usage: nfa regexp string...\n");
		return 1;
	}
	
	post = re2post(argv[1]);
    printf("%s\n", post);
    //return 0;

	if(post == NULL){
		fprintf(stderr, "bad regexp %s\n", argv[1]);
		return 1;
	}

	start = post2nfa(post);
	if(start == NULL){
		fprintf(stderr, "error in post2nfa %s\n", post);
		return 1;
	}
	
	l1.s = malloc(nstate*sizeof l1.s[0]);
	l2.s = malloc(nstate*sizeof l2.s[0]);
	for(i=2; i<argc; i++)
		if(match(start, argv[i]))
			printf("%s\n", argv[i]);
	return 0;
}